Tempo: TraceQL Configurable static fields

adrapereira opened the pull request:
**What is this feature?**

⚠️ This feature is behind the `traceqlSearch` feature toggle. 

This PR turns the previously hardcoded static fields "Service Name" and "Span Name" into a list of fields that can be configured in the datasource configuration page. The defaults are still the same, but now each datasource admin can customise these fields to better serve their specific needs. 

![image](https://user-images.githubusercontent.com/4313971/227255936-1854e59d-0666-4930-8285-bd64319446e5.png)

Users can set the scope, tag, operator and even the value for each field, which will be displayed in the TraceQL Search tab for anyone who starts a new search. Fields with values pre-set will automatically be present in the generated query, which could be handy for instances where users need to regularly search for a specific tag-value combination. 

**Other noteworthy changes:**
- Added more units to the duration field validation and updated the tooltip to show supported units
- Removed static fields from Tags section

**Demo**

https://user-images.githubusercontent.com/4313971/227258853-7a57b327-d28c-448a-a148-f3d507c0a32e.mov




**Why do we need this feature?**

Some users will benefit from easily accessible fields to make their TraceQL search experience faster. 

**Who is this feature for?**

All users who need custom static fields in the TraceQL UI Search. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

adrapereira committed:
TraceQL - configurable static fields for new UI

adrapereira committed:
TraceQL - filter out static fields from Tags section. Added tooltip to static fields

adrapereira committed:
Merge remote-tracking branch 'origin/main' into andre/configurable_traceql_static_fields

adrapereira committed:
Add more units to duration validation. Improve duration field tooltip with accepted units

adrapereira committed:
Better control of delete button on SearchField

adrapereira committed:
Move new config behind feature toggle

adrapereira committed:
Special title for intrinsic "name"

adrapereira committed:
Fix tests

09jvilla commented:
@adrapereira this is awesome! Thanks for addressing everything!

Future follow up idea that came to mind for me is the idea of letting users define a special name for their static field (e.g., Instead of "Span HTTP Status Code", the user could set their own string like "HTTP status code"). But would wait to see if someone asks for this first. Similarly, would wait to see if anyone asks for the ability to force the static field filters to be populated. So not only suggest "hey you have to filter by service.name, but in fact require that at least some value be selected for this filter". The person deciding to do the forcing would be someone who is more like an admin persona who is trying to enforce best practices around searching with their user base. Again, let's not do anything on this unless someone else brings it up. I'm mostly thinking aloud.

09jvilla reviewed (approved):
Left 1 suggestion on the tool tip language.

09jvilla commented on public/app/plugins/datasource/tempo/SearchTraceQLEditor/TraceQLSearch.tsx:
```diff
@@ -85,37 +88,47 @@ const TraceQLSearch = ({ datasource, query, onChange }: Props) => {
+            <InlineSearchField
+              key={f.id}
+              label={filterTitle(f)}
+              tooltip={`Configured in datasource - ${filterScopedTag(f)}`}
```
Not sure if tooltips support links...

aocenas commented on public/app/plugins/datasource/tempo/SearchTraceQLEditor/SearchField.tsx:
```diff
@@ -114,12 +128,16 @@ const SearchField = ({ filter, datasource, updateFilter, deleteFilter, isTagsLoa
           isLoading={isTagsLoading}
-          options={tags.map((t) => ({ label: t, value: t }))}
+          // Add the current tag to the list if it doesn't exist in the tags prop, otherwise the field will be empty even though the state has a value
+          options={(!filter.tag || tags.find((t) => t === filter.tag) ? tags : [filter.tag, ...tags]).map((t) => ({
```
Wouldn't something like `_.uniq([filter.tag, ...tags])` make more sense here? We would not need to do the complicated ternary, and I assume there is not that many value that the perf diff would be noticeable.

aocenas commented on public/app/plugins/datasource/tempo/SearchTraceQLEditor/TraceQLSearch.tsx:
```diff
@@ -85,37 +88,47 @@ const TraceQLSearch = ({ datasource, query, onChange }: Props) => {
+      ?.filter((f) => f.value)
+      .forEach((f) => {
+        if (!findFilter(f.id)) {
+          updateFilter(f);
```
This does not seem right to me. As I understand it we add filters from the config if they are not at the moment in the UI. This makes sense but then we do not remove them when the data source changes. So if I start with one data source this adds it's static fields, then I switch to other one with other static fields I end up with UI that will show static fields from both data sources?

aocenas replied:
Not exactly sure how to fix this but I still get to a state like this if I have 2 data sources with fields setup:
![Screenshot from 2023-03-28 16-28-56](https://user-images.githubusercontent.com/1014802/228271491-9913f2e3-d353-451c-ae26-ebd31204a4d4.png)

I think we just don't clean the query state when changing data source, which may have some value if the data source is the same type so makes sense to keep the query. I am not really sure how the defaults should work in this case. Do we even need the default values?

09jvilla replied:
I like the idea. Use case is "I'm an admin who owns configuration of this datasource and want to save my end users a few clicks by suggesting the value that I anticipate they'll use for filtering in 80%+ of cases." 

That being said, I haven't had any customer feedback one way or another yet. We can ask Paymentsense about this next week when they take this for a drive? I can't tell from this thread how buggy or not you're saying this interaction is - would they be able to test what you're got in this draft PR or are the bugs identified here such that you don't want to merge it at all? Would be nice if they can take it for a drive on their own data and make the determination of if it'd be useful, but if not, I can always ask them about it conceptually. 

Btw, I'd actually be interested to probe how much they find the ability to set default values useful versus the ability to require that the user apply some specific set of filters (i.e. force the user to choose some non-empty value for the static fields). This is more about enforcing good querying practice and less about saving the user clicks/giving them breadcrumbs in-product though.

aocenas replied:
So the main thing is I am not sure how this should work in this particular case.

We preserve query if we move between the same data sources, like switching between Tempo-dev and Tempo-prod. In case of these default fields, should they be preserved? If so should they be merged into the defaults of the other data source somehow? Or override them? Should they be treated as part of the query from the point they are visible or should they be part of the query only if user changes them and only then override/merge with the other data source?

aocenas replied:
@adrapereira I also just now realized that we encode this in the query here: https://github.com/grafana/grafana/blob/bfb0dde4a8b5d297a341ab44846ec9c78b5ea1ef/public/app/plugins/datasource/tempo/dataquery.gen.ts#L91. I am not sure how is this supposed to work if the query should work with different data source? Also if the datasource definition of these fields change should that affect the saved queries in dashboard for example? I think this information should no be encoded in the query as that ties is to the data source config.

aocenas commented on public/app/plugins/datasource/tempo/SearchTraceQLEditor/TraceQLSearch.tsx:
```diff
@@ -85,37 +88,47 @@ const TraceQLSearch = ({ datasource, query, onChange }: Props) => {
     fetchTags();
```
Btw seems like this fetch in case somebody changes the data source before finishing can lead to out of order response and wrong data. Should probably be canceled, or just use `useAsync`.

aocenas commented on public/app/plugins/datasource/tempo/configuration/TraceQLSearchSettings.tsx:
```diff
@@ -0,0 +1,154 @@
+      setDatasource((await dataSourceSrv.get({ type: options.type, uid: options.uid })) as TempoDatasource);
+    };
+    fetchDatasource();
+  }, [dataSourceSrv, options.type, options.uid]);
```
This is not easily solvable and not sure this approach will work. For example you don't know when the URL or auth is set so you may not have a data source that would work. So you could check on the data source what is set, or make the health check request and depending on that start working with it, or just have error handling but that should handle case where it just cannot work like missing URL.

aocenas commented on public/app/plugins/datasource/tempo/configuration/TraceQLSearchSettings.tsx:
```diff
@@ -0,0 +1,154 @@
+        }
+      } catch (error) {
+        if (error instanceof Error) {
+          dispatch(notifyApp(createErrorNotification('Error', error)));
```
In case the data source isn't set up yet we may not want to show an error.

aocenas commented on public/app/plugins/datasource/tempo/configuration/TraceQLSearchSettings.tsx:
```diff
@@ -0,0 +1,154 @@
+  }, [dataSourceSrv, options.type, options.uid]);
+
+  useEffect(() => {
+    const fetchTags = async () => {
```
Again would probably make sense to make this cancellable.

adrapereira committed:
Move static fields not in the datasource to the Tags section

aocenas commented on public/app/plugins/datasource/tempo/SearchTraceQLEditor/SearchField.tsx:
```diff
@@ -30,13 +30,27 @@ interface Props {
   const [isLoadingValues, setIsLoadingValues] = useState(false);
   const [options, setOptions] = useState<Array<SelectableValue<string>>>([]);
-  const [scopedTag, setScopedTag] = useState(scopeHelper(filter) + filter.tag);
+  const [scopedTag, setScopedTag] = useState(filterScopedTag(filter));
```
Not sure I get this. If `filter` changes we add scope and then save it in state if it's new, that triggers rerender which triggers `updateOptions`. If we removed the state and just used `filterScopedTag(filter)` and just memoize it so it returns the same object then only on change of that result would the `updateOptions` be called, but would save additional rerender, isn't that so?

adrapereira committed:
Start using the useAsync hook in the Tempo TraceQL configuration page to retrieve the tags and datasource

adrapereira committed:
Fix tests

adrapereira committed:
Fix test. Use useAsync to retrieve options in SearchField

adrapereira committed:
Remove ability to set a default value in filter configuration.
Removed type from filter, dynamic filters are now any filters not present in the datasource config

adrapereira committed:
Updated the static filters tooltip

adrapereira committed:
Replace useState + useEffect with useMemo for scopedTag
